File: tests/refs/lru_cache.py
"""Reference: LRU cache driven by a list of operations.

The fixture input is `[capacity, ops]` where `ops` is a list of operations:
    ["put", k, v]   -> returns None (not included in output)
    ["get", k]      -> returns the looked-up value (or -1) and appends to output

The driver returns the ordered list of get results.
"""

from __future__ import annotations
from collections import OrderedDict
from typing import List, Any


class _LRU:
    def __init__(self, capacity: int) -> None:
        self.capacity = capacity
        self.cache: "OrderedDict[int, int]" = OrderedDict()

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        self.cache.move_to_end(key)
        return self.cache[key]

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache.move_to_end(key)
        self.cache[key] = value
        if len(self.cache) > self.capacity:
            self.cache.popitem(last=False)


def lruDriver(capacity: int, ops: List[List[Any]]) -> List[int]:
    cache = _LRU(capacity)
    out: List[int] = []
    for op in ops:
        name = op[0]
        if name == "put":
            cache.put(op[1], op[2])
        elif name == "get":
            out.append(cache.get(op[1]))
        else:
            raise ValueError(f"unknown op {name!r}")
    return out

File: tests/refs/LruCache.java
/*
 * Reference Java implementation for tests/cases/lru_cache.json.
 *
 * Driver consumes a list of ops [["put", k, v], ["get", k], ...] over a
 * cache of given capacity and returns the ordered list of get results.
 */
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class LruCache {
    private static class LRU {
        private final int capacity;
        private final LinkedHashMap<Long, Long> map;
        LRU(int capacity) {
            this.capacity = capacity;
            this.map = new LinkedHashMap<>(16, 0.75f, true);
        }
        long get(long key) {
            Long v = map.get(key);
            return v == null ? -1L : v;
        }
        void put(long key, long value) {
            map.put(key, value);
            if (map.size() > capacity) {
                map.remove(map.entrySet().iterator().next().getKey());
            }
        }
    }

    public static List<Long> lruDriver(long capacity, List<List<Object>> ops) {
        LRU cache = new LRU((int) capacity);
        List<Long> out = new ArrayList<>();
        for (List<Object> op : ops) {
            String name = (String) op.get(0);
            if (name.equals("put")) {
                cache.put(((Number) op.get(1)).longValue(),
                          ((Number) op.get(2)).longValue());
            } else if (name.equals("get")) {
                out.add(cache.get(((Number) op.get(1)).longValue()));
            } else {
                throw new IllegalArgumentException("unknown op: " + name);
            }
        }
        return out;
    }
}
